### PR TITLE
add flag to sensor context for first tick after start

### PR DESCRIPTION
**Pull request by prha**
## Summary & Motivation
For sensors that have been turned off for some time, it's useful to have some indication so that user code can decide to limit the range of values to examine (e.g. only look at the last 5 minutes worth of data).

This PR adds the last start time to the stored sensor data, and then adds a flag to the sensor context by comparing the last tick timestamp to the last start timestamp.

## How I Tested These Changes
BK, TBD

### CODE REVIEW

**Comment by prha** [comment]
@gibsondan aside from switching the flag to a timestamp, I forget what the backcompat considerations are for changing these calls...

File: python_modules/dagster/dagster/_api/snapshot_sensor.py [hunk 1]
```diff
@@ -21,6 +21,7 @@ def sync_get_external_sensor_execution_data_ephemeral_grpc(
     last_completion_time: Optional[float],
     last_run_key: Optional[str],
     cursor: Optional[str],
+    last_start_time: Optional[float] = None,
```
Comment (gibsondan) [inline python_modules/dagster/dagster/_api/snapshot_sensor.py:24]: name bikeshedding:

just confirming this is the last time the *sensor* was started right? (as opposed to the last time a tick was started)
maybe worth clarifying that in the name

would "sensor_start_time" be better? Maybe "last" is implied in that case? Or maybe "last_sensor_start_time" or "most_recent_sensor_start_time" is maximally clear

File: python_modules/dagster/dagster/_cli/sensor.py [hunk 2]
```diff
@@ -373,6 +374,8 @@ def execute_cursor_command(sensor_name, cli_args, print_fn):
                             last_run_key=job_state.instigator_data.last_run_key,
                             min_interval=external_sensor.min_interval_seconds,
                             cursor=cursor_value,
+                            last_tick_start_timestamp=job_state.instigator_data.last_tick_start_timestamp,
+                            last_sensor_start_timestamp=job_state.instigator_data.last_sensor_start_timestamp,
```
Comment (gibsondan) [inline python_modules/dagster/dagster/_cli/sensor.py:378]: these names are clearer

File: python_modules/dagster/dagster/_core/definitions/sensor_definition.py [hunk 3]
```diff
@@ -308,6 +311,15 @@ def last_completion_time(self) -> Optional[float]:
         """Optional[float]: Timestamp representing the last time this sensor completed an evaluation."""
         return self._last_completion_time
 
+    @public
+    @property
+    def last_start_time(self) -> Optional[float]:
+        """Optional[float]: Timestamp representing the last time this sensor was started. Can be
```
Comment (gibsondan) [inline python_modules/dagster/dagster/_core/definitions/sensor_definition.py:317]: ah particularly when paired with last_completion_name the symmetry would make me think both times referred to a tick

File: python_modules/dagster/dagster/_core/scheduler/instigation.py [hunk 4]
```diff
@@ -107,6 +110,18 @@ def __new__(
             check.opt_int_param(min_interval, "min_interval"),
             check.opt_str_param(cursor, "cursor"),
             check.opt_float_param(last_tick_start_timestamp, "last_tick_start_timestamp"),
+            check.opt_float_param(last_sensor_start_timestamp, "last_sensor_start_timestamp"),
+        )
+
+    def with_start_time(self, start_time: float) -> "SensorInstigatorData":
```
Comment (gibsondan) [inline python_modules/dagster/dagster/_core/scheduler/instigation.py:116]: with_sensor_start_timestamp

File: python_modules/dagster/dagster_tests/daemon_sensor_tests/test_sensor_run.py [hunk 5]
```diff
@@ -223,6 +223,18 @@ def run_cursor_sensor(context):
     return RunRequest(run_key=None, run_config={}, tags={})
 
 
+@sensor(job_name="the_job")
+def start_skip_sensor(context):
+    # skips the first tick after a start
+    if (
+        not context.last_completion_time
+        or context.last_start_time
+        and context.last_start_time > context.last_completion_time
```
Comment (gibsondan) [inline python_modules/dagster/dagster_tests/daemon_sensor_tests/test_sensor_run.py:232]: testing my knowledge of boolean precedence ordering here (and I think i am failing the test)

Comment (gibsondan) [inline python_modules/dagster/dagster_tests/daemon_sensor_tests/test_sensor_run.py:232]: ok i think i understand the motivation for the boolean now :) Making the change to last_start_time (and maybe renaming last_completion_time to last_tick_completion_time?) will help make this clearer i think

**Review by gibsondan — APPROVED** [review]
some naming suggestions inline to make it maximally clear what is being started

**Pull request merged**